hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/common/IncrementalChunkBuffer.java

```diff
@@ -0,0 +1,217 @@
+  private ByteBuffer getAtIndex(int i) {
+    final ByteBuffer ith = i < buffers.size() ? buffers.get(i) : null;
+    if (ith != null) {
+      Preconditions.checkState(ith.capacity() == increment);
+    }
+    return ith;
+  }
+
+  /** @return the i-th buffer. It may allocate buffers. */
+  private ByteBuffer getAndAllocateAtIndex(int index) {
+    final int limitIndex = limit / increment;
+    // never allocate over limit
+    Preconditions.checkArgument(index <= limitIndex);
+
+    int size = buffers.size();
+    if (index < size) {
+      return getAtIndex(index);
+    }
+
+    // allocate upto the given index
+    ByteBuffer b = null;
+    for (; size <= index; size++) {
+      b = ByteBuffer.allocate(increment);
+      buffers.add(b);
+    }
+    if (index == limitIndex) {
+      b.limit(limit % increment);
+    }
+    return b;
+  }
+
+  /** @return the buffer containing the position. It may allocate buffers. */
+  private ByteBuffer getAndAllocateAtPosition(int position) {
+    final int i = position / increment;
+    final ByteBuffer ith = getAndAllocateAtIndex(i);
+    Preconditions.checkState(ith.position() == position % increment);
+    return ith;
+  }
+
+  @Override
+  public int position() {
+   // The buffers list must be in the following orders:
+   // full buffers, buffer containing the position, empty buffers, null buffers
+    int i = 0;
+    ByteBuffer ith = null;
+    // buffers must be full
+    for (; i < buffers.size(); i++) {
+      ith = getAtIndex(i);
+      if (ith.position() != increment) {
+        break; // found the first non-full buffer
+      }
+    }
+    // buffer containing the position
+    final int position = i * increment + Optional.ofNullable(getAtIndex(i))
+        .map(ByteBuffer::position).orElse(0);
+    assertRemainingList(ith, i);
+    return position;
+  }
+
+  private void assertRemainingList(ByteBuffer ith, int i) {
+    if (ith != null) {
+      // buffers must be empty
+      for (i++; i < buffers.size(); i++) {
+        ith = getAtIndex(i);
+        if (ith == null) {
+          break; // found the first non-null
+        }
+        Preconditions.checkState(ith.position() == 0);
+      }
+    }
+    // buffers must be null
+    for (i++; i < buffers.size(); i++) {
+      Preconditions.checkState(getAtIndex(i) == null);
+    }
+  }
+
+  @Override
+  public int remaining() {
+    return limit - position();
+  }
+
+  @Override
+  public void clear() {
+    buffers.forEach(ByteBuffer::clear);
+  }
+
+  @Override
+  public void put(ByteBuffer that) {
+    if (that.remaining() > this.remaining()) {
+      final BufferOverflowException boe = new BufferOverflowException();
+      boe.initCause(new IllegalArgumentException(
+          "Failed to put since that.remaining() = " + that.remaining()
+              + " > this.remaining() = " + this.remaining()));
+      throw boe;
+    }
+
+    final int thatLimit = that.limit();
+    int p = position();
+    for (; that.position() < thatLimit;) {
```

Comment:
What's the benefit of such a `for` loop vs. a simple `while` loop?

Reply:
I tries to avoid while-loop in general since for-loop is more powerful. :). For example, I may (Indeed, I should) move "int p = position();" inside.

Reply:
> I tries to avoid while-loop in general since for-loop is more powerful. :).

Thanks.

> For example, I may (Indeed, I should) move "int p = position();" inside.

I was considering mentioning this.  Similarly, `p += min` might be more conventional in the last section of the `for` instead of inside the loop body. :)